Added a new section to bit-manipulation.md

ganeshvarbalaji opened the pull request:
Added a new section "Count set bits upto n" under the Useful Tricks section after the Brian Kernighan's algorithm.

The Brian Kernighan's algorithm explains how to count the set bits of an integer n. The new added section explains how to count the set bits of all integers from 1 upto the number n, without giving a TLE during contest submissions.

The Solution was inspired by this [geeksforgeeks problem](https://www.geeksforgeeks.org/problems/count-total-set-bits-1587115620/1)

adamant-pwn commented on src/algebra/bit-manipulation.md:
````diff
@@ -186,6 +186,59 @@ int countSetBits(int n)
+- Subtract $2^x$ from $n$ and Recurse with the algorithm using the new $n$.
+
+```cpp
+int countSetBits(int n){
+	if(n <= 0) return 0;
+	
+	int count = 0;
+	
+	// Find the highest power of 2 that is
+	// less than or equal to the given n.
+	int x = 0;
+	while ((1 << x) <= n) x++;
+	x--;
+	
+	// Count set bits of all the numbers
+	// from 1 to 2^x - 1
+	count = x * (1 << (x - 1));
+	
+	// Count set bits in the most significant
+	// bit of the numbers from 2^x to n.
+	count += (n - (1 << x) + 1);
+	
+	// Recurse for remaining numbers from 2^x to n
+	count += countSetBits(n - (1 << x));
+	
+	return count;
+}
````
```suggestion
int countSetBits(int n) {
        int count = 0;
        while (n > 0) {
            int x = std::bit_width(n) - 1;
            count += x << (x - 1);
            n -= 1 << x;
            count += n + 1;
        }
        return count;
}
```
It probably makes sense to implement the algorithm in a more concise manner, also using `std::bit_width(n)`, and make it iterative, since it's a tail recursion anyway.

adamant-pwn commented on src/algebra/bit-manipulation.md:
```diff
@@ -186,6 +186,59 @@ int countSetBits(int n)
+
+We can see that the all the columns except the leftmost have $4$ (i.e. $2^2$) set bits each, i.e. upto the number $2^3 - 1$, the number of set bits is $3*(2^{3-1})$.
+
+With the new knowledge in hand we can come up with the following algorithm:
+- Find the highest power of $2$ that is lesser than or equal to the given number. Let this number be $x$.
```
```suggestion
With the new knowledge in hand we can come up with the following algorithm:

- Find the highest power of $2$ that is lesser than or equal to the given number. Let this number be $x$.
```
Need a newline, otherwise itemized list misrenders.

adamant-pwn commented on src/algebra/bit-manipulation.md:
```diff
@@ -186,6 +186,59 @@ int countSetBits(int n)
+### Count set bits upto $n$
+To count the number of set bits of all numbers upto the number $n$ (inclusive), we can run the Brian Kernighan's algorithm on all numbers upto $n$. But this will result in a "Time Limit Exceeded" in contest submissions. 
+
+We can use the fact that for numbers upto $2^x$ (i.e. from $1$ to $2^x - 1$) there are $x*2^{x-1}$ set bits. This can be visualised as follows.
```
```suggestion
We can use the fact that for numbers upto $2^x$ (i.e. from $1$ to $2^x - 1$) there are $x \cdot 2^{x-1}$ set bits. This can be visualised as follows.
```

adamant-pwn commented on src/algebra/bit-manipulation.md:
````diff
@@ -186,6 +186,59 @@ int countSetBits(int n)
+8 ->   1 0 0 0
+```
+
+We can see that the all the columns except the leftmost have $4$ (i.e. $2^2$) set bits each, i.e. upto the number $2^3 - 1$, the number of set bits is $3*(2^{3-1})$.
````
```suggestion
We can see that the all the columns except the leftmost have $4$ (i.e. $2^2$) set bits each, i.e. upto the number $2^3 - 1$, the number of set bits is $3 \cdot 2^{3-1}$.
```

adamant-pwn commented on src/algebra/bit-manipulation.md:
```diff
@@ -186,6 +186,59 @@ int countSetBits(int n)
+
+With the new knowledge in hand we can come up with the following algorithm:
+- Find the highest power of $2$ that is lesser than or equal to the given number. Let this number be $x$.
+- Calculate the number of set bits from $1$ to $2^x - 1$ by using the formua $x*(2^{x-1})$.
```
```suggestion
- Calculate the number of set bits from $1$ to $2^x - 1$ by using the formua $x \cdot 2^{x-1}$.
```

adamant-pwn commented on src/algebra/bit-manipulation.md:
```diff
@@ -186,6 +186,59 @@ int countSetBits(int n)
+- Find the highest power of $2$ that is lesser than or equal to the given number. Let this number be $x$.
+- Calculate the number of set bits from $1$ to $2^x - 1$ by using the formua $x*(2^{x-1})$.
+
+The next steps needs more explanation.
+- Count the no. of set bits in the most significant bit from $2^x$ to $n$ and add it.
```
```suggestion
The next steps needs more explanation.

- Count the no. of set bits in the most significant bit from $2^x$ to $n$ and add it.
```
Same here, extra newline is needed for the correct render of itemized list.

adamant-pwn reviewed (changes requested):
Thanks for the suggestion! Do you mind looking through review comments?

ganeshvarbalaji commented:
Hi @adamant-pwn ,

Thanks so much for taking the time to review my pull request! I really appreciate your feedback and suggestions.

I've incorporated your changes into the code and addressed the points you raised. You can find the updated version in the branch now.

I believe everything is ready to go, so please feel free to merge the pull request with the main branch at your convenience.

Thanks again for your help!

adamant-pwn commented:
Thanks!
